packages/components/src/color-list-picker/types.ts

```diff
@@ -0,0 +1,25 @@
+/**
+ * External dependencies
+ */
+import type { CSSProperties } from 'react';
+
+export type ColorListPickerProps = {
+	colors: Array< {
+		name: string;
+		color: NonNullable< CSSProperties[ 'color' ] >;
+	} >;
+	labels: Array< string >;
+	value?: Array< string >;
+	disableCustomColors?: boolean;
+	enableAlpha?: boolean;
+	onChange: ( newValue: Array< string > ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
+};
+
+export type ColorOptionProps = Pick<
+	ColorListPickerProps,
+	'colors' | 'disableCustomColors' | 'enableAlpha'
+> & {
+	label: ColorListPickerProps[ 'labels' ][ number ];
+	value: string; // TODO: can we extract this from ColorListPickerProps['value']? That prop being optional is a little tricky
+	onChange: ( newValue: string ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
```

Comment:
I'm not sure if I fully understand the potential solution you described, but I'm assuming it's something along these lines? If so, yes I think it makes sense!

```suggestion
	value?: Array< string | undefined >;
	disableCustomColors?: boolean;
	enableAlpha?: boolean;
	onChange: ( newValue: Array< string | undefined > ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
};

export type ColorOptionProps = Pick<
	ColorListPickerProps,
	'colors' | 'disableCustomColors' | 'enableAlpha'
> & {
	label: ColorListPickerProps[ 'labels' ][ number ];
	value: string | undefined; // TODO: can we extract this from ColorListPickerProps['value']? That prop being optional is a little tricky
	onChange: ( newValue: string | undefined ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
```

Reply:
Thanks @mirka! I'm getting back into this after being out for the holidays. I did consider the `| undefined` approach, but also (as I tend to do) dove down another rabbit hole at the same time.

The tl;dr of my previously mentioned solution was that the whole reason the `undefined` is needed is that another componenet (`ColorPalette`) explicitly passes an `undefined` value` when de-selecting a color. I think we can use an empty sting instead to both simplify the typing, and prevent a potential edge case bug that TypeScript might not currently see for us, involving an optional parameter that may not be safe to actually omit in all scenarios). More thorough description and a potential change in #46890 😄 

If that's not a good direction though, I'm perfectly happy with your proposed solution of adding `| undefined` as needed!

Reply:
Based on the feedback from @jorgefilipecosta on #46890, I've scrapped that approached and gone with `string | undefined` here.

I've also added some JSDoc descriptions to types. Probably not mission-critical on an internal component like this one but I figured I may as well include them while I was updating things.